### ISSUE TITLE
leoBridge does not work with @auto-md subtrees

### ISSUE THREAD

**Issue by SegundoBob** [leo-editor/leo-editor#258]
When a Leo-Editor file contains one or more @auto-md nodes, leoBridge does NOT see the same Leo-Editor file contents that the user sees by interactively opening the Leo-Editor file.

For all the nodes in a @auto-md subtree, leoBridge sees the node headline "NewHeadline".

Some ways of looping over the nodes using leoBridge yield too few nodes.

For example, I have a Leo-Editor file containing exactly one node, a @auto-md node that reads in a .md file that generates 40 nodes.  My program that uses leoBridge and c.all_positions_iter() to loop over all positions only sees 14 positions.  My program that uses leoBridge and c.all_unique_nodes() to loop over all unique vnodes, correctly sees 41 vnodes.

Test System
```
Xubuntu 64-bit 14.04

Leo Log Window
Leo 5.3-devel, build 20160411083458, Mon Apr 11 08:34:58 CDT 2016
Git repo info: branch = None, commit = 98a461b8cf58
Python 3.4.3, PyQt version 5.2.1
linux
setting leoID from os.getenv('USER'): 'bob'
load dir: /pri/git/leo-editor/leo/core
global config dir: /pri/git/leo-editor/leo/config
home dir: /home/bob
reading settings in /pri/git/leo-editor/leo/config/leoSettings.leo
reading settings in /home/bob/.leo/myLeoSettings.leo
Found /help
reading settings in /pri/bob/1/Leo/Leo.leo
reading: /pri/bob/1/Leo/Leo.leo
read outline in 0.90 seconds
```

**Comment by edreamleo** [comment]
Thanks for the report.  I'll fix it for 5.3-b1

**Comment by edreamleo** [comment]
There may be an import problem in the bridge.  The createImporterData method in leoImport.py finds importers this way:
```
importlib.import_module('leo.plugins.importers.%s' % module_name)
```
If that fails, the @auto node will contain the entire external file. That will explain the different number of nodes in the files. createImporterData issues a warning, but that may be suppressed in the bridge.

**Comment by edreamleo** [comment]
@SegundoBob Please trace the import statement above are report what you see. You may have to write to sys.__stderr__ to make the trace visible.

**Comment by SegundoBob** [comment]
Is this what you want?

My test case test01.leo is a Leo-Editor file containing exactly one node:

@auto-md test01.md

My file test01.md contains 10 level 1 headings and 1 level 2 heading.

Hence, when I interactively open test01.leo without tracing, I see 12 nodes and none of these nodes has headline "NewHeadline". I see that test01.md has been correctly parsed.  I see all the headlines and bodies that I expect.

When I open test01.md using leoBridge without tracing, I see 6 nodes and every node headline is "NewHeadline."

Now when I enable SherlockTracer, interactively opening test01.leo fails to find the markdown importer and all the markdown is put in the body of the only node in the file.  The lines in trace_interactive.txt appear in the log pane.


When I use leoBridge with tracing, my test program sees 6 nodes (not 1 node, not 12 nodes) and the trace is in trace_leoBridge.txt.

The traces are too big for github "issues" to handle, so I sent them to you attached to an email.  I'm including my test case and my tracing code.


Test System
```
Xubuntu 64-bit 14.04
Leo Log Window
Leo 5.3-devel, build 20160411083458, Mon Apr 11 08:34:58 CDT 2016
Git repo info: branch = None, commit = 98a461b8cf58
Python 3.4.3, PyQt version 5.2.1
linux
```

**Comment by edreamleo** [comment]
FYI: The traces contain many references to `<frozen importlib._bootstrap>`. Googling same, leads to [this Python issue](https://bugs.python.org/issue15110). Supposedly, this affects only Python 3.3, but from your report it seems to exist in Python 3.4 as well.

I see you used Sherlock to trace createImporterData. I had intended for you only to trace that one method.  I'll soon up improved tracing code for that method.

If we remove all the cruft, we are left (for the non-bridge case):
```
can not import leo.plugins.importers.javascript
can not import leo.plugins.importers.pascal
can not import leo.plugins.importers.php
can not import leo.plugins.importers.html
...
```
Similarly, the bridge case, with Sherlock tracing enabled, reduces to:
```
SherlockTracer.run:patterns:
+.*
+:.*
__init__.py         .+import_module(name='leo.plugins.impo...)
import_module     level = 0
import_module     if name.startswith('.'):
import_module     return _bootstrap._gcd_import(name[level:], package, level)

can not import leo.plugins.importers.javascript
SherlockTracer.run:patterns:
+.*
+:.*
__init__.py         .+import_module(name='leo.plugins.impo...)
import_module     level = 0
import_module     if name.startswith('.'):
import_module     return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>..+_gcd_import(name='leo.plugins.impo...)

can not import leo.plugins.importers.pascal
```
In short, the imports are failing for you both from the leoBridge and when running Leo normally.  It's a bit of mystery why the failed imports should cause two different symptoms, but it seems that getting the imports to work is the first priority ;-) I shall look into Leo's code relating to sys.path...

**Comment by edreamleo** [comment]
Rev bf0af4c attempts a fix by adding plugins/importers and plugins/writers to the list of modules added to sys.path in adjustSysPath (leoBridge.py only). This is probably needed in any case.

However, the real question for SegundoBob is why none of the @auto importers are being found **outside** the bridge. Indeed, LM.adjustSysPath is a do-nothing, deferring all work to g.importModule, which is complicated, and in any event is _not_ used to find importers.

Usually, everything just works, which perhaps means that people have set up their sys.path to handle it. Or maybe @auto isn't working for many people??

**Comment by edreamleo** [comment]
I just disabled my sitecustomize.py files.  All my tests, including unit tests, still work. So it appears that Leo is able to find @auto importers & writers out of the box.

Rev ead33c8 fixes a crasher in the nodetags.py plugin when running the unit test:
`@test leoBridge init logic`
This test must be run externally.

In short, everything appears to be working properly.  I'll leave the bug open for comments for a bit longer.

**Comment by SegundoBob** [comment]
# Leo-Editor Interactive: Python 3 @auto-md now broken, but Python 2 still works

You have broken @auto-md for Leo-Editor interactive using Python 3, but it still works using Python 2.

Now using Python 3 test01.leo puts all of test01.md into the body of the @auto-md node.  There are no errors in the Log pane and no errors to the console.

Now using Python 2 test01.leo has 12 nodes when it is opened by Leo-Editor.  That is, the markdown parser creates 11 nodes.

# test01.py - command line options added

I enhanced test01.py with more command line options. A few minutes ago, I sent it to you attached to an email.

# leoBridge.controller().openLeoFile() is messed up

Command line
```
python3 test01.py -s -b -p test01
```
shows that instead of opening test01.leo, leoBridge tries to open t.leo, e.leo, s.leo, t.leo, 0.leo, 1.leo.  That is, bridge.openLeoFile("test01.leo") causes "test01" to be split into individual letters.  Perhaps, the change to allow wildcards on the command line is at fault.  I tried passing openLeoFile() a list of filenames, but that didn't work either.
```
$ python3 test01.py  -b -p test01
global config directory : /pri/git/leo-editor/leo/config
home directory : /home/bob
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/t.leo. creating new window
import-jupyter-notebook requires nbformat package
NewHeadline bob.20160419092730.1
wrote recent file: /home/bob/.leo/.leoRecentFiles.txt
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/e.leo. creating new window
NewHeadline bob.20160419092730.2
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/s.leo. creating new window
NewHeadline bob.20160419092730.3
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/t.leo. creating new window
NewHeadline bob.20160419092730.4
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/0.leo. creating new window
NewHeadline bob.20160419092730.5
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/1.leo. creating new window
NewHeadline bob.20160419092730.6
2016-04-19 09:27:30 /pri/git/leo_bug_demos/leoBridge_markdown
```

# 3 Minor Bugs

Now command line
```
python3 test01.py -s -b test01
```
shows that even when leoBridge is run with "silent=True" and "verbose=True", the following three messages are printed on the console:
```
global config directory : /pri/git/leo-editor/leo/config
home directory : /home/bob
import-jupyter-notebook requires nbformat package
```
These are bugs.  They should not be printed.  A "silent" guard is needed around these three prints.


# Test System
```
Xubuntu 64-bit 14.04

Leo Log Window
Leo 5.3-devel, build 20160412153848, Tue Apr 12 15:38:48 CDT 2016
Git repo info: branch = None, commit = a59b49f8639b
Python 3.4.3, PyQt version 5.2.1
linux
```
and
```
Leo Log Window
Leo 5.3-devel, build 20160412153848, Tue Apr 12 15:38:48 CDT 2016
Git repo info: branch = None, commit = a59b49f8639b
Python 2.7.6, PyQt version 4.8.6
linux2
```

**Comment by SegundoBob** [comment]
# Leo-Editor Interactive: @auto-md now works with both Python 2 and Python 3

I did not change my revision of Leo-Editor.  I don't know what changed.  I only know that when I tried to investigate why @auto-md was failing with Python 3, it had stopped failing.

**Comment by edreamleo** [comment]
@SegundoBob Thanks for the report.  bf0af4c8 (two days ago) added plugins/importers and plugins/writers to sys.path in some situations.  It's possible that may have fixed things.

But I doubt it. There are several unit tests that will fail if the @auto importers or writers can't be found. Such tests have not failed recently, indicating to me that @auto has, in fact, been working all along.

I'll leave this issue open while I fix the not-quite-silent problem. In some situations, this could be very annoying.  Furthermore, I've made a note to create a better test bed for the Leo bridge on my end, so it's all good :-)

**Comment by SegundoBob** [comment]
I hope you noticed the most important problem:

# leoBridge.controller().openLeoFile() is messed up

Command line
```
python3 test01.py -s -b -p test01
```
shows that instead of opening test01.leo, leoBridge tries to open t.leo, e.leo, s.leo, t.leo, 0.leo, 1.leo. That is, bridge.openLeoFile("test01.leo") causes "test01" to be split into individual letters. Perhaps, the change to allow wildcards on the command line is at fault. I tried passing openLeoFile() a list of filenames, but that didn't work either.
```
$ python3 test01.py  -b -p test01
global config directory : /pri/git/leo-editor/leo/config
home directory : /home/bob
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/t.leo. creating new window
import-jupyter-notebook requires nbformat package
NewHeadline bob.20160419092730.1
wrote recent file: /home/bob/.leo/.leoRecentFiles.txt
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/e.leo. creating new window
NewHeadline bob.20160419092730.2
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/s.leo. creating new window
NewHeadline bob.20160419092730.3
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/t.leo. creating new window
NewHeadline bob.20160419092730.4
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/0.leo. creating new window
NewHeadline bob.20160419092730.5
file not found: /pri/git/leo_bug_demos/leoBridge_markdown/1.leo. creating new window
NewHeadline bob.20160419092730.6
2016-04-19 09:27:30 /pri/git/leo_bug_demos/leoBridge_markdown
```

**Comment by edreamleo** [comment]
> instead of opening test01.leo, leoBridge tries to open t.leo, e.leo, s.leo, t.leo, 0.leo, 1.leo.

Oops!

**Comment by edreamleo** [comment]
a6f7e5be, cf9d9d3 and f97a27b fix this part of this issue:

> when leoBridge is run with "silent=True" and "verbose=True", the following three messages are printed on the console:

```
global config directory : /pri/git/leo-editor/leo/config
home directory : /home/bob
import-jupyter-notebook requires nbformat package
```

- a6f7e5be: silent=True now suppresses the first two lines. They come from leoBridge.py itself.
- cf9d9d3: importers/ipynb.py no longer issues the warning in the last line.
- f97a27b: `leo-bridge-test.py` shows how to suppress all messages from g.es_print.

I don't recommend killing g.es_print, but the choice is yours.

**Comment by edreamleo** [comment]
> instead of opening test01.leo, leoBridge tries to open t.leo, e.leo, s.leo, t.leo, 0.leo, 1.leo.

leoBridge has nothing to do with this.  test01.leo contains the argument-handling code in question.

See LM.scanOptions and especially its helper, LM.computeFilesList, for an example of how to parse the command line.

Imo, this closes this issue.

**Comment by SegundoBob** [comment]
I'm sorry for the bug in test01.py that caused me to **mistakenly** think there was a problem with leoBridge.controller().openLeoFile() .

I verified the 3 "silent" fixes in
```
** isPython3: True
Leo 5.3-devel, build 20160412153848, Tue Apr 12 15:38:48 CDT 2016
Git repo info: branch = None, commit = ec44e9a0c88c
Python 3.4.3, LeoGui: dummy version
linux
```
I agree with you:  this issue is closed.

**Comment by edreamleo** [comment]
On Thu, Apr 21, 2016 at 3:47 PM, SegundoBob <notifications@github.com>
wrote:

> I'm sorry for the bug in test01.py that caused me to *mistakenly* think
> there was a problem with leoBridge.controller().openLeoFile() .
>
No problem.  Good things came of this mistake.​

> I verified the 3 "silent" fixes in
>
> ** isPython3: True
> Leo 5.3-devel, build 20160412153848, Tue Apr 12 15:38:48 CDT 2016
> Git repo info: branch = None, commit = ec44e9a0c88c
> Python 3.4.3, LeoGui: dummy version
> linux
>
> I agree with you: this issue is closed.
>
​Thanks for the confirmation.

Edward
